lib/matplotlib/widgets.py

```diff
@@ -2595,3 +2598,190 @@ def onmove(self, event):
+        y0, y1 = self.ax.get_ylim()
+        if y0 < y1:
+            origin = 'lower'
+        else:
+            origin = 'upper'
+        props = dict(cmap=self.cmap, alpha=0.5, origin=origin,
+                     norm=NoNorm(), visible=False, zorder=1,
+                     extent=(x0, x1, y0, y1), aspect=self.ax.get_aspect())
+        props.update(overlay_props or {})
+
+        self._resize(None)
+        self._overlay_plot = self.ax.imshow(self._overlay, **props)
+        self.artists = [self._cursor, self._overlay_plot]
+        self.connect_event('resize_event', self._resize)
+
+        # These must be called last
+        self.label = 1
+        self.radius = radius
+        self._drawing = True
+        for artist in self.artists:
+            artist.set_visible(True)
+
+    @property
+    def overlay(self):
+        """The paint overlay image"""
+        return self._overlay
+
+    @overlay.setter
+    def overlay(self, image):
+        self._overlay = image
+        if image is None:
+            self.ax.images.remove(self._overlay_plot)
+            self.update()
+            return
+        self._overlay_plot.set_data(image)
+        self._shape = image.shape
+        x0, x1 = self.ax.get_xlim()
+        y0, y1 = self.ax.get_ylim()
+        self._overlay_plot.set_extent((x0, x1, y0, y1))
+        # Update the radii and window.
+        self.radius = self._radius
+        self.update()
+
+    @property
+    def label(self):
+        """The label index"""
+        return self._label
+
+    @label.setter
+    def label(self, value):
+        if value >= self.cmap.N:
+            raise ValueError('Maximum label value = %s' % len(self.cmap - 1))
+        self._label = value
+        self._cursor.set_edgecolor(self.cmap(value))
+
+    @property
+    def radius(self):
+        """The label radius in pixels"""
+        return self._radius
+
+    @radius.setter
+    def radius(self, r):
+        self._radius = r
+        xfm = self.ax.transData.inverted()
+        x0, y0 = xfm.transform((0, 0))
+        x1, y1 = xfm.transform((r, r))
+        self._rx, self._ry = abs(x1 - x0), abs(y1 - y0)
+
+        self._cursor.set_width(self._rx * 2)
+        self._cursor.set_height(self._ry * 2)
+
+    def _press(self, event):
+        self._update_cursor(event.xdata, event.ydata)
+        self._update_overlay(event.x, event.y)
+        self.update()
+
+    def _onmove(self, event):
+        self._update_cursor(event.xdata, event.ydata)
+        if event.button and event.button in self.validButtons:
+            self._update_overlay(event.x, event.y)
+        self.update()
+        self.onselect(event.xdata, event.ydata)
+
+    def _resize(self, event):
+        extents = self.ax.get_window_extent().extents
+        self._offsetx = extents[0]
+        self._offsety = extents[1]
+        self._shape = (int(extents[3] - extents[1]),
+                       int(extents[2] - extents[0]))
+        self._overlay = np.zeros(self._shape, dtype='uint8')
```

Comment:
You may be able to use `_image.resample` in the same way [that `imshow` eventually does](https://github.com/matplotlib/matplotlib/blob/master/lib/matplotlib/image.py#L407) if you want take a stab at preserving the overlay.